agent.go

```diff
@@ -251,13 +259,25 @@ func (agent *agentS) fullRequestResponse(url string, method string, data interfa
 		j, err = json.Marshal(data)
 	}
 
+	var cancel context.CancelFunc
+	if ctx == nil {
+		ctx, cancel = context.WithTimeout(context.Background(), agent.clientTimeout)
+		defer cancel()
+	}
+	if _, ok := ctx.Deadline(); !ok {
```

Comment:
I'm not sure I follow why client timeout is only used if there is no deadline set. Is this to not to override longer deadline? In this case I'd argue this is the right place to add this timeout to the context. Since it's the caller to decide which period of time to use, I'd suggest we set the timeout there.

Reply:
The idea was to ensure that in the case when the function is called with, for instance: "context.Background", set Deadline here. If ctx has already Deadline, keep it.

If the caller decided to set a specific Deadline it will be used.

Reply:
Alright, so what's the purpose of setting the timeout in this method? My current understanding is that this is some kind of default that should satisfy any use case, but I'd argue that this should be the biggest value out of announce and data transmission values, and not the smallest one, i.e. as per your other comment, there should be no problem with sending data using 15s timeout, but an announcement limited to 5s will likely to fail. So I'd suggest to use 15s as a default instead. This would also simplify the implementation here, since the shortest timeout will cancel the context even if the longer one has not been reached yet, meaning that you can always do
```go
ctx, cancel := context.WithTimeout(ctx, agent.announceTimeout)
```
and let the data transmission methods set their own shorter deadline.

Reply:
I believe this would make the logic clearer too: by default agent client gives up after 15s, but for data transmission requests we require shorter timeouts to not to overload the agent.

Reply:
I think you are right that code is less straightforward now 😃 . @andrewslotin please take a look at https://github.com/instana/go-sensor/pull/280/commits/0e369bc1fd05aa309579a30d09436e8eba0b2e7e. I tried to simplify and make it clearer. So now:

1. Client has a timeout (which is equal to announceTimeout which is the biggest we need so far). Before it was 5s and now it is a 15s.
2. We have 3 methods that call actual call (`fullRequestResponse` method):

```
func (agent *agentS) request(url string, method string, data interface{}) (string, error) {
	ctx, cancel := context.WithTimeout(context.Background(), agent.clientTimeout)
	defer cancel()
	return agent.fullRequestResponse(ctx, url, method, data, nil, "")
}

func (agent *agentS) announceRequest(url string, method string, data interface{}, ret *agentResponse) (string, error) {
	return agent.fullRequestResponse(context.Background(), url, method, data, ret, "")
}

func (agent *agentS) requestHeader(url string, method string, header string) (string, error) {
	ctx, cancel := context.WithTimeout(context.Background(), agent.clientTimeout)
	defer cancel()
	return agent.fullRequestResponse(ctx, url, method, nil, nil, header)
}
```

2 of them overwrite (reduce) timeout, so their behavior will be as before. And announce call will use a default one that is equal to 15s now.